server/api/services/transaction_service.go

```diff
@@ -0,0 +1,176 @@
+package services
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/go-webauthn/webauthn/protocol"
+	"github.com/go-webauthn/webauthn/webauthn"
+	"github.com/labstack/echo/v4"
+	"github.com/teamhanko/passkey-server/api/dto/intern"
+	"github.com/teamhanko/passkey-server/persistence/models"
+	"github.com/teamhanko/passkey-server/persistence/persisters"
+	"net/http"
+)
+
+type TransactionService interface {
+	Initialize(userId string, transaction *models.Transaction) (*protocol.CredentialAssertion, error)
+	Finalize(req *protocol.ParsedCredentialAssertionData) (string, string, *models.Transaction, error)
+}
+
+type TransactionServiceCreateParams struct {
+	*WebauthnServiceCreateParams
+
+	TransactionPersister persisters.TransactionPersister
+}
+
+type transactionService struct {
+	*WebauthnService
+
+	transactionPersister persisters.TransactionPersister
+}
+
+func NewTransactionService(params TransactionServiceCreateParams) TransactionService {
+	return &transactionService{
+		WebauthnService: &WebauthnService{
+			BaseService: &BaseService{
+				logger:              params.Ctx.Logger(),
+				tenant:              params.Tenant,
+				credentialPersister: params.CredentialPersister,
+			},
+
+			webauthnClient: params.WebauthnClient,
+			generator:      params.Generator,
+
+			userPersister:        params.UserPersister,
+			sessionDataPersister: params.SessionPersister,
+		},
+		transactionPersister: params.TransactionPersister,
+	}
+}
+
+func (ts *transactionService) Initialize(userId string, transaction *models.Transaction) (*protocol.CredentialAssertion, error) {
+	webauthnUser, err := ts.userPersister.GetByUserId(userId, ts.tenant.ID)
+	if err != nil {
+		ts.logger.Error(err)
+		return nil, echo.NewHTTPError(http.StatusNotFound, "Unable to find user")
+	}
+
+	if webauthnUser == nil {
+		return nil, echo.NewHTTPError(http.StatusNotFound, "Unable to find user")
+	}
+
+	credentialAssertion, sessionData, err := ts.webauthnClient.BeginLogin(
+		intern.NewWebauthnUser(*webauthnUser),
+		webauthn.WithUserVerification(ts.tenant.Config.WebauthnConfig.UserVerification),
+		ts.withTransaction(transaction.Identifier, transaction.Data),
+	)
+	if err != nil {
+		ts.logger.Error(err)
+		return nil, echo.NewHTTPError(
+			http.StatusBadRequest,
```

Comment:
I think we shouldn't return a `BadRequest` status code here, instead we should return `InternalServerError`.

Reply:
Hmm that one was tricky because BeginLogin also returns a bad request error when the user has no suitable credentials for logging in. I added a check for that before BeginLogin to reflect both error cases